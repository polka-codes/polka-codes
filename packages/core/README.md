# Polka Codes Core

[![npm version](https://img.shields.io/npm/v/@polka-codes/core.svg)](https://www.npmjs.com/package/@polka-codes/core)
[![npm downloads](https://img.shields.io/npm/dm/@polka-codes/core.svg)](https://www.npmjs.com/package/@polka-codes/core)
[![License](https://img.shields.io/npm/l/@polka-codes/core.svg)](https://github.com/polkacodes/polkacodes/blob/main/LICENSE)

Core AI services and agent implementations for Polka Codes framework.

## Features

- Multiple AI provider support (Ollama, Anthropic, DeepSeek)
- Extensible agent architecture
- Tool integration system
- Type-safe API
- Logging and monitoring

## Installation

```bash
bun add @polka-codes/core
```

## Usage

### Basic Setup

```typescript
import { AgentBase } from '@polka-codes/core';

class MyAgent extends AgentBase {
  // Implement your agent logic
}

const agent = new MyAgent();
await agent.runTask('your-task');
```

### Available Services

- `AnthropicService`: Integration with Anthropic API
- `DeepSeekService`: Integration with DeepSeek API
- `OllamaService`: Integration with Ollama API

### Tools

The core package provides a tool system for extending functionality:

```typescript
import { ToolBase } from '@polka-codes/core';

class MyTool extends ToolBase {
  // Implement your tool logic
}
```

## Development

### Building

```bash
cd packages/core
bun run build
```

### Testing

```bash
bun test
```

---

*This README was generated by polka.codes*
