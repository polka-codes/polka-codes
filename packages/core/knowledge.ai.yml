files:
  "src/Agent/index.ts":
    description: "Base agent implementation and agent registry"
    api:
      classes:
        1:
          name: "AgentBase"
          methods:
            1:
              name: "run"
              params:
                1: { name: "task", type: "string" }
              returns: "Promise<void>"
  "src/Agent/MultiAgent.ts":
    description: "Multi-agent coordination system"
    api:
      classes:
        1:
          name: "MultiAgent"
          methods:
            1:
              name: "delegateTask"
              params:
                1: { name: "task", type: "string" }
                2: { name: "context", type: "object" }
              returns: "Promise<void>"
  "src/AiService/index.ts":
    description: "AI service abstraction layer"
    api:
      classes:
        1:
          name: "AiServiceBase"
          methods:
            1:
              name: "complete"
              params:
                1: { name: "prompt", type: "string" }
              returns: "Promise<string>"
  "src/tools/index.ts":
    description: "Tool definitions and registry"
    api:
      functions:
        1:
          name: "getAvailableTools"
          params: []
          returns: "Tool[]"
  "src/tools/updateKnowledge.ts":
    description: "Tool for updating knowledge.ai.yml files with smart merging capabilities"
    api:
      functions:
        1:
          name: "updateKnowledge"
          description: "Updates knowledge files with smart merging of YAML content"
          params:
            1: { name: "path", type: "string", description: "Directory path where knowledge file is located" }
            2: { name: "file", type: "string", description: "Optional specific file path" }
            3: { name: "knowledge", type: "string", description: "YAML content to merge" }
          returns: "Promise<ToolResponse>"

invariants:
  1: "All agents must extend AgentBase"
  2: "AI services must implement AiServiceBase"
  3: "Tools must follow the Tool interface"

patterns:
  1: "Agent implementations use prompt templates"
  2: "AI services follow a provider pattern"
  3: "Tools use XML-style formatting for execution"

rules:
  1: "Each agent must have corresponding prompt templates"
  2: "All tools must have tests"
  3: "AI service calls must be metered"
  4: "Knowledge management policies must be followed"
  5: "Use numbered dictionary format instead of arrays in knowledge files"
