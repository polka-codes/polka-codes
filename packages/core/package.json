{
  "name": "@polka-codes/core",
  "version": "0.8.17",
  "license": "AGPL-3.0",
  "author": "github@polka.codes",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": ["dist"],
  "scripts": {
    "build": "tsup src/index.ts --experimental-dts --format esm --clean"
  },
  "dependencies": {
    "@anthropic-ai/sdk": "0.39.0",
    "lodash": "^4.17.21",
    "openai": "^4.80.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.2"
  }
}
