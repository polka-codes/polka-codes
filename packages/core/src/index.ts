export * from './AiService'
export * from './Agent'
export * from './tools'
export * from './AiService/ModelInfo'
export * from './AiTool'
export * from './tool'
export * from './getAvailableTools'
export * from './config'
export type { ToolFormat } from './config'
export * from './tools/utils/getArg'
export { replaceInFile as replaceInFileHelper } from './tools/utils/replaceInFile'
