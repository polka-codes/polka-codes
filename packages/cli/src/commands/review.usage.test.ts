// Unit test to verify review command usage consolidation behavior
// generated by polka.codes

import { describe, expect, test } from 'bun:test'
import { UsageMeter } from '@polka-codes/core'

describe('Review command usage consolidation', () => {
  test('should consolidate usage tracking across review workflow and task execution', () => {
    // Simulate the usage meter being shared between review workflow and task execution
    const sharedUsageMeter = new UsageMeter({
      anthropic: {
        'claude-3-5-sonnet-20241022': {
          inputPrice: 3,
          outputPrice: 15,
          cacheWritesPrice: 3.75,
          cacheReadsPrice: 0.3,
        },
      },
    })

    // Simulate review workflow usage
    sharedUsageMeter.incrementMessageCount(1)
    const reviewUsage = sharedUsageMeter.usage

    // Simulate task execution using the same usage meter
    sharedUsageMeter.incrementMessageCount(2)
    const finalUsage = sharedUsageMeter.usage

    // Verify that the usage is cumulative
    expect(finalUsage.messageCount).toBe(3) // 1 from review + 2 from task
    expect(finalUsage.messageCount).toBeGreaterThan(reviewUsage.messageCount)

    // Verify usage consolidation prevents double printing
    expect(finalUsage).toMatchSnapshot()
  })

  test('should handle case where no task runs after review', () => {
    const usageMeter = new UsageMeter()

    // Simulate review workflow only
    usageMeter.incrementMessageCount(1)
    const usage = usageMeter.usage

    expect(usage.messageCount).toBe(1)
    expect(usage).toMatchSnapshot()
  })
})
