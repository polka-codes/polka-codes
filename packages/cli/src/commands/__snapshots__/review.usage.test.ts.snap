// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Review command usage consolidation should consolidate usage tracking across review workflow and task execution 1`] = `
{
  "cachedRead": 0,
  "cost": 0,
  "input": 0,
  "messageCount": 3,
  "output": 0,
}
`;

exports[`Review command usage consolidation should handle case where no task runs after review 1`] = `
{
  "cachedRead": 0,
  "cost": 0,
  "input": 0,
  "messageCount": 1,
  "output": 0,
}
`;
